Run truffle from anywhere

mrice32 opened the pull request:
This PR includes a few simplifications:
- All tests from all packages are run using `npx lerna run --stream --concurrency=1 test` (this is what `yarn run test` will run at root). Note: Ganache is still required to be up on port 9545. Stream shows the output from the test commands as they come in (otherwise, it appears as if it's hanging) and concurrency keeps it from trying to run all the tests in parallel. I haven't done extensive testing with concurrency not set, so it may work, but I wanted to be careful before blindly suggesting it since there may be issues with nonces in Ganache.
- ci has been modified to do all testing through the above command. The (mostly useless) voter dapp test was set up so it was covered in this command as well, so this job was removed from ci as a simplification.
- `yarn run truffle *` can be run within any package.
- `yarn run test` can be used within any package to run the tests (Ganache still required for most).
- Buidler was only being used to test core. This is still the case. It is likely possible to set buidler up to work similarly to truffle in each package, but that is out of the scope of this PR.
- This should allow the CLI tool to be easily moved out into its own package with few modifications.
- `lerna` detects a cyclic dependency in the repo between financial-templates-lib and core. Moving the CLI tool out will help with this. However, there are two scripts in core that depend on financial-templates-lib. We should either move these particular scripts elsewhere or move _all_ scripts in core to some other package. This is an open question. Just as a matter of simplicity, it's probably easier to just deal with those two scripts and leave the rest as they are unless there's another reason to move them. Definitely interested to hear thoughts on this.

EDIT (next to bullets):
- `yarn run truffle *` is runnable at root. The only exception is that `yarn run truffle test` will not find any tests since none are configured at root.
- `yarn run load-addresses` will work at root and in `core`. This will run `apply-registry` to load the mainnet addresses into the contracts. Open question as to whether this should be configured to be runnable in all packages. For now, I only added it to root and core.

This PR is a solution to running truffle from within each package individually. The reason it was done this way was to allow the other packages to continue to depend on the migrations defined in `core` for testing. As explained below, if a package ceases to need those migrations or wants to implement its own migrations, there is a mutually exclusive setup that will also work.

### Packages that rely on `core` migrations (as all of the packages that use truffle test do)

This requires a specialized truffle config that effectively resolves the `@umaprotocol/core` dependency directory and then tells truffle to use the location of that package as its context for finding contracts, migrations, and build artifacts. While I don't think there's anything wrong with this solution, it essentially requires that each package use the exact same contracts, migrations, and artifacts that core does. This is fine as the packages stand today, and it probably makes sense as a simplifying measure.

Note: this setup allows artifacts.require to work _exactly_ as if running from inside core. Contracts can be referenced as so: `artifacts.require("SomeContract")`.

### Packages that use their own (or no) migrations

If a package wants to have a unique set of migrations that are more simplified than `core` or no migrations at all, it requires a mutually exclusive setup (that would work equally well). The major change would be that this package would remove all of the specialized options from the truffle config and reference artifacts as so: `artifacts.require("@umaprotocol/core/SomeContract")` throughout the tests and scripts.

mrice32 commented:
@chrismaree @nicholaspai could y'all take a look when you get a chance?

chrismaree commented on disputer/truffle-config.js:
```diff
@@ -0,0 +1,9 @@
+const path = require("path");
```
this is a pretty cool solution +1

nicholaspai commented on disputer/truffle-config.js:
```diff
@@ -0,0 +1,9 @@
+const wkdir = path.dirname(require.resolve("@umaprotocol/core/package.json"));
+
+module.exports = {
+  ...require("@umaprotocol/common").TruffleConfig,
```
Interesting.

nicholaspai commented on financial-templates-lib/truffle-config.js:
```diff
@@ -0,0 +1,9 @@
+const path = require("path");
```
Since this `config` is repeated for all of the packages, is it worth writing it in a single file ("truffle-config-template.js") and importing that from all the packages?

mrice32 commented:
@chrismaree @nicholaspai the reason this config is repeated is that I'm trying to avoid another circular dependency (we already have one or two that we'll need to eliminate. It's a little subtle, and I should've explained.

If you add the config in `umaprotocol/common`, it needs to be able to `require` `umaprotocol/core` to resolve the directory of the truffle context. However, `umaprotocol/common` is imported by `umaprotocol/core`, so this would mean that core and common depend on each other, which gives a warning in lerna and probably won't work once the packages are put up on npm (because they'll recursively install one another I think). So, there are two choices that I can think of that are relatively simple:
- Add multiple configs as is done now.
- Instead of `TruffleConfig` being exported by `common`, we export a function that returns `TruffleConfig`. The function takes a single argument, and that argument is the directory truffle should use for its context. This would mean the truffle configs would look like this:
```js
const path = require("path");
const wkdir = path.dirname(require.resolve("@umaprotocol/core/package.json"));
const { getTruffleConfig } = require("@umaprotocol/common");
module.exports = getTruffleConfig(wkdir);
```

Thoughts?

mrice32 commented:
> This is great! our packages are still very heavily dependent on `core`. If I was to install a single package from npm, say the `disputer`, would it be possible to run the tests from that package? My understanding is you would only be able to do that if you had the full mono-repo pulled down, right?

Nope, that should be possible (although, I'm not sure tests get installed with npm packages, generally)! The package depends on `umaprotocol/core`, so that package will be installed as a sub-dependency. The require resolution in these config files is built to work seamlessly in the npm single package installation context. The contracts/artifacts in core will be used at runtime and in tests, and the migrations will only be used in tests.

mrice32 commented:
@nicholaspai 

> * Is there a way to remove the repeated `truffle-config.js` code?

See the options listed above.

> * If I want to run truffle from a package, how would I replicate `$(npm bin)` with `yarn`. In other words, does `yarn run truffle` use the truffle version installed in `node_modules`?

Yes, I believe that `yarn run truffle` uses the truffle version in `node_modules`. However, I think `$(npm bin)/truffle` should work within each of these directories too if you still prefer using that form. I think, in general, `yarn run` is a drop-in replacement for `$(npm bin)/X` _and_ `npm run` (i.e. it runs package scripts too).

nicholaspai commented:
> Instead of TruffleConfig being exported by common, we export a function that returns TruffleConfig. The function takes a single argument, and that argument is the directory truffle should use for its context. This would mean the truffle configs would look like this:

@mrice32 @chrismaree I like this solution because you only write the actual config contents once.

mrice32 commented:
> LGTM after you consider implementing the one `truffle-config` refactor I commented about!

Sounds good. I'll implement that!

mrice32 commented:
@chrismaree @nicholaspai I've implemented y'all's suggestion to import as much of that config as possible from elsewhere. PTAL.

I'm adding this to the description, but as a part of this change, I also added a truffle config to root as well, so most truffle commands should work from root (albeit `yarn run truffle test` will find no tests since there's no test directory at root, this should be run using `yarn run test` instead.).

I also added a `load-addresses` command to both `core` and root so you can call `yarn run load-addresses` to load the deployed addresses into the contracts.

nicholaspai commented:
> I'm adding this to the description, but as a part of this change, I also added a truffle config to root as well, so most truffle commands should work from root (albeit yarn run truffle test will find no tests since there's no test directory at root, this should be run using yarn run test instead.).

This is going to be very useful very nice.

> I also added a load-addresses command to both core and root so you can call yarn run load-addresses to load the deployed addresses into the contracts.

How does this differ from `apply-registry`

nicholaspai commented on core/truffle-config.js:
```diff
@@ -1 +1 @@
-module.exports = require("@umaprotocol/common").TruffleConfig;
+module.exports = require("@umaprotocol/common").getTruffleConfig(__dirname);
```
This is a really nice way to call it while passing in a relative path

nicholaspai commented on disputer/truffle-config.js:
```diff
@@ -0,0 +1,4 @@
+const path = require("path");
+const wkdir = path.dirname(require.resolve("@umaprotocol/core/package.json"));
```
What does `require.resolve` do?

mrice32 replied:
I really like the way they describe it in the nodejs docs: "Uses the internal require() machinery to look up the location of a module, but rather than loading the module, just returns the resolved filename."

mrice32 replied:
To put it differently, it basically tells you the exact path to a module rather than reading the module.

Side note: the reason `package.json` is used is that it's a file that should always exist within all modules and isn't dependent on a specific entrypoint or index file existing.

mrice32 commented:
> How does this differ from `apply-registry`

It runs apply registry. I figured `load-addresses` was a little more intuitive, and I needed to name it something different to get yarn to differentiate it from the `apply-registry` binary.

chrismaree commented on truffle-config.js:
```diff
@@ -0,0 +1,2 @@
+// Note: this config is only here as a matter of convenience to allow all truffle commands that can be run from core work at root as well.
+module.exports = require("@umaprotocol/common").getTruffleConfig("./core");
```
this is the only instance where you specify a direct path because it will always be executed from the `root`, right? could you have used the `const path = require("path");` syntax here otherwise?

mrice32 replied:
This is a good point. This is because root is never going to be a public package and it technically doesn't (and maybe can't?) depend on one of its sub-packages. I think that syntax might work, but it really shouldn't because `@umaprotocol/core` isn't a dependency of root.

mrice32 commented:
> I have one quick question: do we plan on refactoring the repo directory structure to follow more standard nesting? For example, do we want to make a `packages/*` folder to store all the packages? clearly, this is a separate PR but just wanted to ask about that here.

As discussed IRL, I think we should do this, but do it after everything else is lerna-ized and working, so it's basically just a file move and not much more.
